include/eve/module/algo/algo/keep_if.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/module/core.hpp>
+#include <eve/module/algo/algo/array_utils.hpp>
+#include <eve/module/algo/algo/concepts.hpp>
+#include <eve/module/algo/algo/common_forceinline_lambdas.hpp>
+#include <eve/module/algo/algo/for_each_iteration.hpp>
+#include <eve/module/algo/algo/preprocess_range.hpp>
+#include <eve/module/algo/algo/traits.hpp>
+
+#include <array>
+
+namespace eve::algo
+{
+  template <typename TraitsSupport>
+  struct keep_if_ : TraitsSupport
+  {
+    template <typename UnalignedI, typename P>
+    struct delegate
+    {
+      explicit delegate(UnalignedI out, P p) : out(out), p(p) {}
+
+      EVE_FORCEINLINE bool step(auto it, eve::relative_conditional_expr auto ignore, auto /*idx*/)
+      {
+        auto loaded  = eve::load[ignore](it);
+        auto mask    = p(loaded);
+        auto density = density_for_compress_copy<typename TraitsSupport::traits_type>();
+        out          = compress_copy[unsafe][density][ignore](it, loaded, mask, out);
+        return false;
+      }
+
+      template <typename I, std::size_t size>
+      EVE_FORCEINLINE bool unrolled_step(std::array<I, size> arr)
+      {
+        array_map(arr, call_single_step(this));
+        return false;
+      }
+
+      UnalignedI out;
+      P p;
+    };
+
+    template <relaxed_range Rng, typename P>
+    EVE_FORCEINLINE auto operator()(Rng&& rng, P p) const
+    {
+      if (rng.begin() == rng.end()) return unalign(rng.begin());
+
+      auto processed = preprocess_range(TraitsSupport::get_traits(), EVE_FWD(rng));
+
+      auto iteration = algo::for_each_iteration(processed.traits(), processed.begin(), processed.end());
+      auto out = iteration.base;
+      delegate<unaligned_t<decltype(out)>, P> d{unalign(out), p};
+      iteration(d);
+      return unalign(rng.begin()) + (d.out - processed.begin());
+    }
+  };
+
+  //================================================================================================
+  //! @addtogroup algos
+  //! @{
+  //!   @var keep_if
+  //!   @brief In-place version of `eve::algo::copy_if`
+  //!
+  //!   **Defined in Header**
+  //!
+  //!   @code
+  //!   #include <eve/module/algo.hpp>
+  //!   @endcode
+  //!
+  //!   Takes values that pass the predicate, and copies them to the beginning of the range.
+  //!   Values past the returned sentinel are to be considered garbage.
+  //!
+  //!   Acts like an in-place version of `::copy_if`, or an eager version of filter.
+  //!
+  //!   @note
+  //!   If you need to apply a transformation, you can use `eve::algo::views::map` or `eve::algo::transform_keep_if`.
+  //!
+  //!   **Tuning**
+  //!
+  //!   `::dense_output`/`::sparse_output`: if you expect very sparse output (one or two `true`s per register),
```

Comment:
What's with global namespace? i bet it doesn't work. Bug?

Reply:
the leading `::` is to force doxygen to create links. it doesn't show up on the rendered documents.

Reply:
OK - let's have a look when it's rendered.

Reply:
try eve::dense_output. 
Does dense_output even has a doxygen entry ?

Reply:
it does, on my machine writing `::dense_output` and rendering creates a link to the correct (very bare) page

Reply:
![image](https://github.com/jfalcou/eve/assets/56922769/8fba7379-4eae-4db9-b273-b4bf79addbed)

![image](https://github.com/jfalcou/eve/assets/56922769/6cab3c1d-b435-42a4-be37-cf67fb2b258e)